plutus-core-interpreter/src/Language/PlutusCore/Interpreter/LMachine.hs

```diff
@@ -0,0 +1,278 @@
+    , EvaluationResult
+    , evaluateL
+    , runL
+    ) where
+
+import           Language.PlutusCore
+import           Language.PlutusCore.Constant
+import           Language.PlutusCore.View
+import           PlutusPrelude
+
+import           Control.Monad.Identity
+import           Data.IntMap                  (IntMap)
+import qualified Data.IntMap                  as IntMap
+
+type Plain f = f TyName Name ()
+
+-- Let's just throw exceptions for the time being: error handling is changeable at the moment
+data LMachineException =
+    LMachineException String (Plain Term)
+  | LMachineStringException String
+    deriving (Show, Typeable)
+
+instance Exception LMachineException
+
+-- | A term together with an enviroment mapping free variables to heap locations
+data Closure = Closure
+    { _closureValue       :: Plain Term
+    , _closureEnvironment :: Environment
+    } deriving (Show)
+
+-- | L machine environments
+-- Each entry is a mapping from the 'Unique' representing a variable to a heap location
+newtype Environment = Environment (IntMap HeapLoc)
+    deriving (Show)
+
+-- | Heap location. Int should always be big enough: maxBound::Int is 2^63-1
+type HeapLoc = Int
+
+-- | Heap entries: unevaluated and evaluated lazy function arguments.
+data HeapEntry =
+    Evaluated   Closure  -- The term in this closure should always be a value. It would be good if the type system could distiniguish these.
+  | Unevaluated Closure
+
+-- | The heap itself: a map from heap locations to
+-- unevaluated/evaluated terms, together with an integer counter
+-- @_top@ which serves as the current heap location.
+data Heap = Heap {
+      _heap :: IntMap HeapEntry
+    , _top  :: Int
+}
+
+{- Environments map variable ids to locations, the heap maps locations
+   to terms.  We really do require this indirection: there may be
+   multiple objects in the heap corresponding to the same variable
+   (during recursion, for example).  For the same reason we always
+   work with closures instead of terms; this also makes it easier
+   to ensure that we're using the correct environment. -}
+
+
+-- | Stack frames.  These are effectively continuations: would using explicit CPS speed things up?
+-- The spec also has a frame for n-ary built in application, but that's not in the AST yet
+data Frame
+    = FrameAppArg Closure                        -- ^ @[_ N]@       -- Tells the machine that we've got back to N after evaluating M in [M N]
+    | FrameHeapUpdate HeapLoc                    -- ^ @(update l)@  -- Mark the stack for evaluation of a delayed argument
+    | FrameTyInstArg (Type TyName ())            -- ^ @{_ A}@
+    | FrameUnwrap                                -- ^ @(unwrap _)@
+    | FrameWrap () (TyName ()) (Type TyName ())  -- ^ @(wrap α A _)@
+      deriving (Show)
+
+-- | A context is a stack of frames.
+type EvaluationContext = [Frame]
+
+-- | A local version of the EvaluationResult type using closures instead of values
+data LMachineResult
```

Comment:
This seems a shame - can we make `MachineResult` parametric or something?

Reply:
It's already, just in a weird non-exported way. See `Language.PlutusCore.Evaluation.Result`:

```haskell
data EvaluationResultF a
    = EvaluationSuccess a
    | EvaluationFailure

type EvaluationResult = EvaluationResultF (Value TyName Name ())
```

I have no objections to changing that to

```haskell
data EvaluationResult a
    = EvaluationSuccess a
    | EvaluationFailure

type EvaluationResultDef = EvaluationResult (Value TyName Name ())
```

Reply:
I'll do that, but maybe in a separate pull request for clarity.